Python/ceval.c

```diff
@@ -3610,6 +3866,215 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
+                PUSH(Py_True);
+                DISPATCH();
+            }
+            // Bail on some possible Sequences that we intentionally exclude:
+            if (PyType_FastSubclass(Py_TYPE(subject),
+                                    Py_TPFLAGS_BYTES_SUBCLASS |
+                                    Py_TPFLAGS_UNICODE_SUBCLASS) ||
+                PyByteArray_Check(subject))
+            {
+                Py_INCREF(Py_False);
+                PUSH(Py_False);
+                DISPATCH();
+            }
+            // Lazily import _collections_abc.Sequence, and keep it handy on the
+            // PyInterpreterState struct (it gets cleaned up at exit):
+            PyInterpreterState *interp = PyInterpreterState_Get();
+            if (!interp->seq_abc) {
+                PyObject *abc = PyImport_ImportModule("_collections_abc");
+                if (!abc) {
+                    goto error;
+                }
+                interp->seq_abc = PyObject_GetAttrString(abc, "Sequence");
+                if (!interp->seq_abc) {
+                    goto error;
+                }
+            }
+            int match = PyObject_IsInstance(subject, interp->seq_abc);
+            if (match < 0) {
+                goto error;
+            }
+            PUSH(PyBool_FromLong(match));
+            DISPATCH();
+        }
+
+        case TARGET(MATCH_KEYS): {
+            // Final stage of mapping pattern matches (after Mapping and length
+            // checks): extract the desired values for further destructuring.
+            PyObject *keys = TOP();
+            PyObject *subject = SECOND();
+            assert(PyTuple_CheckExact(keys));
+            PyObject *values = match_keys(tstate, subject, keys);
+            if (!values) {
+                if (_PyErr_Occurred(tstate)) {
+                    goto error;
+                }
+                Py_INCREF(Py_False);
+                PUSH(Py_False);
+                DISPATCH();
+            }
+            // Success!
+            if (oparg) {
+                // If oparg is nonzero, collect the remaining items into a dict
+                // and put it on the stack where the subject used to be:
+                PyObject *rest = dict_without_keys(tstate, subject, keys);
+                if (!rest) {
+                    goto error;
+                }
+                assert(PyDict_CheckExact(rest));
+                SET_SECOND(rest);
+                Py_DECREF(subject);
+            }
+            SET_TOP(values);
+            Py_DECREF(keys);
+            Py_INCREF(Py_True);
+            PUSH(Py_True);
+            DISPATCH();
+        }
+
+        case TARGET(GET_INDEX): {
+            // PUSH(TOS[oparg])
+            PyObject *item = PySequence_GetItem(TOP(), oparg);
+            if (!item) {
+                goto error;
+            }
+            PUSH(item);
+            DISPATCH();
+        }
+
+        case TARGET(GET_INDEX_END): {
+            // PUSH(TOS[TOS1 - 1 - oparg])
```

Comment:
This bytecode seems unnecessary.
What is wrong with?
```
ROT_TWO
LOAD_CONST  oparg+1
BINARY_SUBTRACT
BINARY_SUBSCRIPT
```

Reply:
I originally added the `GET_INDEX` and `GET_INDEX_END` opcodes to try to avoid unnecessary conversions between `PyLongObject*` and `Py_ssize_t`. As you point out, though, it's probably a nano-optimization that's not worth the maintenance burden.

I think I agree that `BINARY_SUBCR` is a fine replacement for both right now.